pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,364 @@
+      Serializer<V> valueSerializer) {
+
+    producerConfig = configs;
+
+    project = configs.getString(ProducerConfig.PROJECT_CONFIG);
+
+    if (keySerializer == null) {
+      this.keySerializer = configs.getConfiguredInstance(
+          ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+      this.keySerializer.configure(configs.originals(), true);
+    } else {
+      configs.ignore(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+      this.keySerializer = keySerializer;
+    }
+
+    if (valueSerializer == null) {
+      this.valueSerializer = configs.getConfiguredInstance(
+          ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+      this.valueSerializer.configure(configs.originals(), false);
+    } else {
+      configs.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+      this.valueSerializer = valueSerializer;
+    }
+
+    lock = new Semaphore(1, true);
+    closed = new AtomicBoolean(false);
+    retries = configs.getInt(ProducerConfig.RETRIES_CONFIG);
+    retriesMs = configs.getLong(ProducerConfig.RETRY_BACKOFF_MS_CONFIG);
+    lingerMs = configs.getLong(ProducerConfig.LINGER_MS_CONFIG);
+    batchSize = configs.getInt(ProducerConfig.BATCH_SIZE_CONFIG);
+    autoCreate = configs.getBoolean(ProducerConfig.AUTO_CREATE_CONFIG);
+    elementCount = configs.getLong(ProducerConfig.ELEMENTS_COUNT_CONFIG);
+    isAcks = configs.getString(ProducerConfig.ACKS_CONFIG).matches("1|all");
+
+    publishers = Collections.synchronizedMap(new HashMap<String, Publisher>());
+  }
+
+  private Publisher createPublisher(String topic) {
+    Publisher pub = null;
+    TopicName topicName = null;
+    TopicAdminClient topicAdmin = null;
+
+    try {
+      topicName = TopicName.create(project, topic);
+      topicAdmin = TopicAdminClient.create();
+      topicAdmin.getTopic(topicName);
+    } catch (Exception e) {
+      if (autoCreate)
+        topicAdmin.createTopic(topicName);
+      else {
+        throw new KafkaException("Failed to construct kafka producer, Topic not found.", e);
+      }
+    }
+
+    try {
+      pub = Publisher.defaultBuilder(topicName)
+          .setBatchingSettings(BatchingSettings.newBuilder()
+              .setElementCountThreshold(elementCount)
+              .setRequestByteThreshold((long) batchSize)
+              .setDelayThreshold(Duration.ofMillis(lingerMs))
+              .build())
+          .setRetrySettings(RetrySettings.newBuilder()
+              .setMaxAttempts(retries)
+              .setRetryDelayMultiplier(MULTIPLIER)
+              .setInitialRetryDelay(Duration.ofMillis(retriesMs))
+              .setMaxRetryDelay(Duration.ofMillis((retries + 1) * retriesMs))
+              .setRpcTimeoutMultiplier(MULTIPLIER)
+              .setInitialRpcTimeout(Duration.ofSeconds(TOTAL_DELAY))
+              .setMaxRpcTimeout(Duration.ofSeconds(TOTAL_DELAY))
+              .setTotalTimeout(Duration.ofSeconds(TOTAL_DELAY))
```

Comment:
There is a semantic difference concerning requests and batches.
I will use it for now, but there maybe a better way to manipulate it.